### PR TITLE
Windows linker cleanup

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

File: crates/linker/src/lib.rs [hunk 1]
```diff
@@ -110,9 +107,9 @@ pub fn build_and_preprocess_host(
 
     preprocess(
         target,
-        dynhost.to_str().unwrap(),
-        metadata.to_str().unwrap(),
-        preprocessed_host_path.to_str().unwrap(),
+        &dynhost,
+        &metadata,
+        preprocessed_host_path,
```
Comment (folkertdev) [inline crates/linker/src/lib.rs:112]: not sure why paths were turned into strings here?! Path captures the usage better I think

File: crates/linker/src/lib.rs [hunk 2]
```diff
@@ -923,11 +936,11 @@ pub fn preprocess(
 fn gen_macho_le(
     exec_data: &[u8],
     md: &mut metadata::Metadata,
-    out_filename: &str,
+    out_filename: &Path,
     macho_load_so_offset: usize,
     _target: &Triple,
     _verbose: bool,
-) -> (MmapMut, File) {
```
Comment (folkertdev) [inline crates/linker/src/lib.rs:930]: we don't need to hold onto the file

File: crates/linker/src/pe.rs [hunk 3]
```diff
@@ -58,81 +62,112 @@ struct PeMetadata {
...
+            Ok(data) => data,
+            Err(err) => {
+                internal_error!("Failed to deserialize metadata: {}", err);
+            }
         }
-    };
+    }
+
+    fn from_preprocessed_host(preprocessed_data: &[u8], new_sections: &[[u8; 8]]) -> Self {
+        use object::ObjectSection;
```
Comment (folkertdev) [inline crates/linker/src/pe.rs:87]: gave metadata collection its own function

File: crates/linker/src/lib.rs [hunk 4]
```diff
@@ -473,25 +460,60 @@ pub fn preprocess(
...
+                host_exe_path,
+                metadata_path,
+                preprocessed_path,
+                verbose,
+                time,
+            )
+            .unwrap_or_else(|e| internal_error!("{}", e));
+        }
+
+        target_lexicon::BinaryFormat::Wasm => {
```
Comment (bhansconnect) [inline crates/linker/src/lib.rs:487]: This shouldn't actually come up? They would use the linker for wasm that brian wrote? Or is that not the case if we use the llvm backend?

File: crates/linker/src/lib.rs (hunk 1 above)
Comment (bhansconnect) [reply crates/linker/src/lib.rs:112]: No particular reason. Your change sounds better.

File: crates/linker/src/lib.rs (hunk 2 above)
Comment (bhansconnect) [reply crates/linker/src/lib.rs:930]: Interesting. I guess I just missed that in some refactor that pulled this function out.

**Review by bhansconnect — APPROVED** [review]
Looks great!

File: crates/linker/src/lib.rs (hunk 2 above)
Comment (folkertdev) [reply crates/linker/src/lib.rs:930]: I think I did this originally, thinking that not dropping the file was relevant. But it isn't, so we can get rid of it

**Pull request merged**